### PR TITLE
Add Quantization Gemm feature

### PR DESCRIPTION
**Pull request by swathiananth**
Changes to add Quantize Gemm metacommand feature

### CODE REVIEW

**Comment by swathiananth** [comment]
The changes are not yet complete. But would like your feedback.

File: tools/common_lib/src/gemm.h [hunk 1]
```diff
@@ -351,6 +354,108 @@ class Gemm : public DirectMlBaseNode
             auto gemm_inp_b = dml::Reinterpret(reshaped_split, dml::TensorDimensions{ batch, head_count, seq, head_size }, input_1_strides);
             outputs_[0] = dml::Gemm(input_0_, gemm_inp_b, dml::NullOpt, DML_MATRIX_TRANSFORM_NONE, DML_MATRIX_TRANSFORM_NONE, alpha, beta, fused_act);
         }
+        else if (type_ == GemmType::GemmType_Quant)
```
Comment (smarcink) [inline tools/common_lib/src/gemm.h:357]: As written above - if you extend this class than no need for another type, and `GemmType::GemmType_AB` condition could be modified.

File: tools/common_lib/src/gemm.h [hunk 2]
```diff
@@ -713,13 +969,197 @@ class GemmBaseDispatcher : public NodeDispatcher
...
+    
+
+    GemmBaseDispatcher(create_params_t&& params, ID3D12Device* d3d12_device, IDMLDevice* dml_device, IDMLCommandRecorder* dml_cmd_recorder, ID3D12GraphicsCommandList* cmd_list)
+        : params_(std::move(params))
+        , dml_cmd_recorder_(dml_cmd_recorder)
+        , dml_device_(dml_device)
+        , d3d12_device_(d3d12_device)
+
+    {
+        if (params_.type == GemmType::GemmType_Quant)
```
Comment (smarcink) [inline tools/common_lib/src/gemm.h:998]: This is really convoluted design. We have condition here which calls some ctor, which inside calls some DML base class which again is checking if it's quantized gemm. 
Please keep it simple and just create fully separated dispatcher for qGEMM or keep extending ordinary "AB" gemm.

File: tools/common_lib/src/gemm.h [hunk 3]
```diff
@@ -376,6 +481,63 @@ class Gemm : public DirectMlBaseNode
         create_operator_impl();
     }
 
+    void record_execute(IDMLCommandRecorder* dml_cmd_recorder, ID3D12GraphicsCommandList* cmd_list,
```
Comment (smarcink) [inline tools/common_lib/src/gemm.h:484]: just keep single `record_execute` and pass nullptr for resoucres which are not used

File: tools/common_lib/src/gemm.h [hunk 4]
```diff
@@ -424,6 +586,69 @@ class Gemm : public DirectMlBaseNode
         record_execute_impl(dml_cmd_recorder, cmd_list, input_bindings, output_bindings);
     }
 
+    virtual void record_initialize(IDMLCommandRecorder* dml_cmd_recorder, ID3D12GraphicsCommandList* cmd_list, ID3D12Resource* resource_b, ID3D12Resource* resource_c,
```
Comment (smarcink) [inline tools/common_lib/src/gemm.h:589]: just keep single `record_initialize` and pass nullptr for resources which are not used

File: tools/common_lib/src/layers_utils.h [hunk 5]
```diff
@@ -351,6 +351,67 @@ inline void fill_with_constant_linear_container_half(std::span<std::byte> contai
...
+{
+    if (value < min) return min;
+    else if (value > max) return max;
+    else return value;
+}
+
+// This function is not taking care of unaligned sizes
+inline void fill_quantized_data_half_to_uint4(std::span<std::byte> output_data, std::span<std::byte> input_data, uint32_t chunk_size, std::span<std::byte> output_scale, std::span<std::byte> output_zeropoint)
+{
+    using Dt = Half;
```
Comment (smarcink) [inline tools/common_lib/src/layers_utils.h:364]: is this `using Dt = Half;` really needed? below you are using `Half` anyway

File: tools/common_lib/src/layers_utils.h [hunk 6]
```diff
@@ -351,6 +351,67 @@ inline void fill_with_constant_linear_container_half(std::span<std::byte> contai
...
+        Half InputMax = 0;
+        Half InputMin = 0xFFFF;
+        uint8_t min = 0;
+        uint8_t max = limit;
+
+        // Find the max and min elements in the chunk
+        for (auto i = chunk_id * chunk_size; i < (chunk_id + 1) * chunk_size; i++)
+        {
+            if (input_ptr[i] > InputMax)
+                InputMax = input_ptr[i];
```
Comment (smarcink) [inline tools/common_lib/src/layers_utils.h:380]: we are using bracelets 
`{ 
}`
even for one line code blocks

File: tools/common_lib/src/layers_utils.h [hunk 7]
```diff
@@ -351,6 +351,67 @@ inline void fill_with_constant_linear_container_half(std::span<std::byte> contai
...
+    else return value;
+}
+
+// This function is not taking care of unaligned sizes
+inline void fill_quantized_data_half_to_uint4(std::span<std::byte> output_data, std::span<std::byte> input_data, uint32_t chunk_size, std::span<std::byte> output_scale, std::span<std::byte> output_zeropoint)
+{
+    using Dt = Half;
+    auto limit = (2 << 3) - 1;
+
+    auto* input_ptr = reinterpret_cast<Dt*>(input_data.data());
```
Comment (smarcink) [inline tools/common_lib/src/layers_utils.h:367]: const auto, we dont want to modify input data by accident

File: tools/common_lib/src/tensor_shape.h [hunk 8]
```diff
@@ -46,6 +46,11 @@ struct TensorShape
 
         return ret;
     }
+
+    inline std::uint32_t get_element_count() const
+    {
+        return (n * c * h * w);
```
Comment (smarcink) [inline tools/common_lib/src/tensor_shape.h:52]: what if some dimension is equal to 0? please have a look at function above `get_dims_count()` and implement this in similar fashion

**Review by smarcink — CHANGES_REQUESTED** [review]
Left comment about convoluted conditions which will make this code harder to maintain (that's why I mark this PR as request changes). Besides that added few minor comments.

File: tools/common_lib/src/gemm.h [hunk 9]
```diff
@@ -635,42 +838,73 @@ class GemmBaseDispatcher : public NodeDispatcher
...
         else if (params_.dt == DataType::eFp16)
         {
             randomize_linear_container_half(random_generator, uniform_distribution, input_data_a_);
-            randomize_linear_container_half(random_generator, uniform_distribution, input_data_b_);
-            randomize_linear_container_half(random_generator, uniform_distribution, input_data_c_);
+            if (params_.shape_c.get_dims_count() > 0)
+            {
+                randomize_linear_container_half(random_generator, uniform_distribution, input_data_c_);
+            }
+            randomize_linear_container_half(random_generator, uniform_distribution, input_data_dequantized_b_);
```
Comment (smarcink) [inline tools/common_lib/src/gemm.h:854]: We could use call this function with different range of random values for different parts of the buffer, just to add some more data variance to test quantization a bit better.

File: tools/common_lib/src/gemm.h [hunk 10]
```diff
@@ -27,6 +27,9 @@ enum class GemmType
     // q + kv
     GemmType_QK_Q_KV,
     GemmType_SV_S_KV,
+
+    // Quantized Gemm
+    GemmType_Quant
```
[the remark this answers is not in the document]
Comment (swathiananth) [reply tools/common_lib/src/gemm.h]: Have restructured the code and made Quant Gemm to be a separate class.

File: tools/common_lib/src/gemm.h [hunk 11]
```diff
@@ -351,6 +354,108 @@ class Gemm : public DirectMlBaseNode
...
+                dml::TensorDesc::Dimensions dimensions_1;
+                dimensions_1.push_back(shape_b.n);
+                dimensions_1.push_back(shape_b.c);
+                dimensions_1.push_back(b_transposed ? shape_b.w : shape_b.h); // to make cmd arguments have normal matrix with transpose internal to each dispatcher
+                dimensions_1.push_back(b_transposed ? shape_b.h : shape_b.w); // to make cmd arguments have normal matrix with transpose internal to each dispatcher
+                dml::TensorDesc desc_input_1 = { quantizedDataType, b_managed ? DML_TENSOR_FLAG_OWNED_BY_DML : DML_TENSOR_FLAG_NONE, dimensions_1 };
+                input_1_ = dml::InputTensor(graph_, 1, desc_input_1);
+
+        
+                uint32_t index = 2;
```
[the remark this answers is not in the document]
Comment (swathiananth) [reply tools/common_lib/src/gemm.h]: Done

File: tools/common_lib/src/gemm.h (hunk 1 above)
Comment (swathiananth) [reply tools/common_lib/src/gemm.h]: Have restructured the code and made Quant Gemm to be a separate class.

File: tools/common_lib/src/gemm.h (hunk 2 above)
Comment (swathiananth) [reply tools/common_lib/src/gemm.h]: Have restructured the code and made Quant Gemm to be a separate class.

File: tools/common_lib/src/gemm.h (hunk 3 above)
Comment (swathiananth) [reply tools/common_lib/src/gemm.h]: Have restructured the code and made Quant Gemm to be a separate class.

File: tools/common_lib/src/gemm.h (hunk 4 above)
Comment (swathiananth) [reply tools/common_lib/src/gemm.h]: Have restructured the code and made Quant Gemm to be a separate class.

File: tools/common_lib/src/gemm.h [hunk 12]
```diff
@@ -351,6 +354,108 @@ class Gemm : public DirectMlBaseNode
             auto gemm_inp_b = dml::Reinterpret(reshaped_split, dml::TensorDimensions{ batch, head_count, seq, head_size }, input_1_strides);
             outputs_[0] = dml::Gemm(input_0_, gemm_inp_b, dml::NullOpt, DML_MATRIX_TRANSFORM_NONE, DML_MATRIX_TRANSFORM_NONE, alpha, beta, fused_act);
         }
+        else if (type_ == GemmType::GemmType_Quant)
+        {
+            bool A_is_quantized = false;
```
[the remark this answers is not in the document]
Comment (swathiananth) [reply tools/common_lib/src/gemm.h]: Doesn't apply after restructure

File: tools/common_lib/src/tensor_shape.h (hunk 8 above)
Comment (swathiananth) [reply tools/common_lib/src/tensor_shape.h]: Done

File: tools/common_lib/src/layers_utils.h (hunk 7 above)
Comment (swathiananth) [reply tools/common_lib/src/layers_utils.h]: Done

File: tools/common_lib/src/layers_utils.h (hunk 6 above)
Comment (swathiananth) [reply tools/common_lib/src/layers_utils.h:384]: Done

File: tools/common_lib/src/layers_utils.h (hunk 5 above)
Comment (swathiananth) [reply tools/common_lib/src/layers_utils.h]: Done

File: tools/common_lib/src/layers_utils.h [hunk 13]
```diff
@@ -36,15 +36,18 @@ enum class DataType
...
 
-inline std::uint8_t get_data_type_bytes_width(DataType dt)
+//inline std::uint8_t get_data_type_bytes_width(DataType dt)
+inline float get_data_type_bytes_width(DataType dt)
 {
     switch (dt)
     {
     case DataType::eFp32: return sizeof(float);
     case DataType::eFp16: return sizeof(std::uint16_t);
+    case DataType::eUint4: return 0.5;
```
Comment (smarcink) [inline tools/common_lib/src/layers_utils.h:50]: Should be "0.5f", because without "f" literal this indirect cast from double to float

Comment (smarcink) [reply tools/common_lib/src/layers_utils.h:50]: also please remove commented line above

File: tools/common_lib/src/layers_utils.h [hunk 14]
```diff
@@ -351,11 +355,75 @@ inline void fill_with_constant_linear_container_half(std::span<std::byte> contai
     }
 }
 
+inline uint8_t clamp_value(uint8_t value, uint8_t min, uint8_t max)
+{
+    if (value < min) return min;
```
Comment (smarcink) [inline tools/common_lib/src/layers_utils.h:360]: Wouldn't be std::clamp(...) enough? Doc: https://en.cppreference.com/w/cpp/algorithm/clamp

File: tools/common_lib/src/conv.h [hunk 15]
```diff
@@ -1506,7 +1506,7 @@ class ConvolutionCmDispatcher : public ConvolutionBaseDispatcher
                     const std::uint32_t ic_chunks_per_hw_thread = 8;
                     const std::uint32_t exec_size = 8;
                     const std::uint32_t dpas_depth = 8;
-                    const std::uint32_t out_dt_size = get_data_type_bytes_width(output_dt);
+                    const std::uint32_t out_dt_size = (std::uint8_t)get_data_type_bytes_width(output_dt);
```
Comment (smarcink) [inline tools/common_lib/src/conv.h:1509]: Variable is uint32_t, but cast to uint8_t was added, why not make these data types the same?

File: tools/common_lib/src/gemm.h [hunk 16]
```diff
@@ -424,77 +424,818 @@ class Gemm : public DirectMlBaseNode
...
+
+        // populate the gemm input tensors here
+        const auto B = get_batch();
+        const auto C = get_channels();
+        const auto M = get_M();
+        const auto K = get_K();
+        const auto N = get_N();
+        std::cout << std::format("Running [B, C, M, K, N]: [{}, {}, {}, {}, {}]\n", B, C, M, K, N);
+
+#if 1
```
Comment (smarcink) [inline tools/common_lib/src/gemm.h:863]: `#if 1`  do we need this?

File: tools/common_lib/src/tensor_shape.h [hunk 17]
```diff
@@ -46,6 +46,33 @@ struct TensorShape
 
         return ret;
     }
+
+    inline std::uint32_t get_element_count() const
+    {
+        std::uint32_t ret = 0;
```
Comment (smarcink) [inline tools/common_lib/src/tensor_shape.h:52]: Function could be much simplified, by i.e.:
1) Initialize `ret` with `n`
2) create lambda functions and reuse it for c, h, w dimensions.

**Review by smarcink — APPROVED** [review]
Approving, left minor comments, please address them and feel free to merge

File: tools/common_lib/src/layers_utils.h (hunk 13 above)
Comment (swathiananth) [reply tools/common_lib/src/layers_utils.h:50]: Done

File: tools/common_lib/src/conv.h (hunk 15 above)
Comment (swathiananth) [reply tools/common_lib/src/conv.h:1509]: Changed

File: tools/common_lib/src/layers_utils.h (hunk 14 above)
Comment (swathiananth) [reply tools/common_lib/src/layers_utils.h:360]: Changed

File: tools/common_lib/src/gemm.h (hunk 9 above)
Comment (swathiananth) [reply tools/common_lib/src/gemm.h:885]: Will do this change later as a different PR.

File: tools/common_lib/src/tensor_shape.h (hunk 17 above)
Comment (swathiananth) [reply tools/common_lib/src/tensor_shape.h]: Done

**Pull request merged**